From b3be317b981f98e43c1d4c5a37be0f6dd1180f4f Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Sun, 23 Apr 2023 21:53:08 +0900
Subject: [PATCH] 1

---
 json_util.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/json_util.c b/json_util.c
index 3e6a6c6..f16d708 100644
--- a/json_util.c
+++ b/json_util.c
@@ -60,7 +60,7 @@ static int _json_object_to_fd(int fd, struct json_object *obj, int flags, const
 
 static char _last_err[256] = "";
 
-const char *json_util_get_last_err()
+const char *json_util_get_last_err(void)
 {
 	if (_last_err[0] == '\0')
 		return NULL;
-- 
2.17.1

