From fbc235f91eea1bf82b77938a12a0bb98b1e35e96 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Sun, 10 Mar 2024 22:49:09 -0400
Subject: [PATCH] fix __text_gen_insn function error

---
 arch/x86/kernel/paravirt.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kernel/paravirt.c b/arch/x86/kernel/paravirt.c
index 5bea8d9..09a1c8e 100644
--- a/arch/x86/kernel/paravirt.c
+++ b/arch/x86/kernel/paravirt.c
@@ -31,7 +31,7 @@
 #include <asm/special_insns.h>
 #include <asm/tlb.h>
 #include <asm/io_bitmap.h>
-
+#include <asm/text-patching.h>
 /*
  * nop stub, which must not clobber anything *including the stack* to
  * avoid confusing the entry prologues.
-- 
2.17.1

