From c9a75f8c9e4e19b03080ba980e7fd765e23f9f55 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Fri, 6 Jan 2023 17:39:02 +0800
Subject: [PATCH] 1

---
 src/rfc3315.c | 58 ++++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 48 insertions(+), 10 deletions(-)

diff --git a/src/rfc3315.c b/src/rfc3315.c
index 8754481..63928bc 100644
--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -48,7 +48,7 @@ static int build_ia(struct state *state, int *t1cntr);
 static void end_ia(int t1cntr, unsigned int min_time, int do_fuzz);
 static void mark_context_used(struct state *state, struct in6_addr *addr);
 static void mark_config_used(struct dhcp_context *context, struct in6_addr *addr);
-static int check_address(struct state *state, struct in6_addr *addr);
+static int check_address(struct state *state, struct dhcp_config *config, struct in6_addr *addr);
 static int config_valid(struct dhcp_config *config, struct dhcp_context *context, struct in6_addr *addr, struct state *state, time_t now);
 static struct addrlist *config_implies(struct dhcp_config *config, struct dhcp_context *context, struct in6_addr *addr);
 static void add_address(struct state *state, struct dhcp_context *context, unsigned int lease_time, void *ia_option, 
@@ -689,9 +689,14 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 		      }
 		    else if (!(c = address6_available(state->context, &req_addr, solicit_tags, plain_range)))
 		      continue; /* not an address we're allowed */
-		    else if (!check_address(state, &req_addr))
+		    else if (!check_address(state, config, &req_addr))
 		      continue; /* address leased elsewhere */
-		    
+		    else if (state->mac_len && config &&
+			     config_has_mac(config, state->mac, state->mac_len, state->mac_type) &&
+			     match_netid(c->filter, solicit_tags, plain_range) &&
+			     !config_implies(config, c, &req_addr))
+		      continue; /* another static address is configured */
+
 		    /* add address to output packet */
 		    add_address(state, c, lease_time, ia_option, &min_time, &req_addr, now);
 		    mark_context_used(state, &req_addr);
@@ -702,7 +707,10 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 	    
 	    /* Suggest configured address(es) */
 	    for (c = state->context; c; c = c->current) 
-	      if (!(c->flags & CONTEXT_CONF_USED) &&
+	      if ((!(c->flags & CONTEXT_CONF_USED) ||
+		   (state->mac_len && config &&
+		    config_has_mac(config, state->mac, state->mac_len, state->mac_type)
+		   )) &&
 		  match_netid(c->filter, solicit_tags, plain_range) &&
 		  config_valid(config, c, &addr, state, now))
 		{
@@ -726,6 +734,12 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 		req_addr = ltmp->addr6;
 		if ((c = address6_available(state->context, &req_addr, solicit_tags, plain_range)))
 		  {
+		    if (state->mac_len && config &&
+			config_has_mac(config, state->mac, state->mac_len, state->mac_type) &&
+			match_netid(c->filter, solicit_tags, plain_range) &&
+			!config_implies(config, c, &req_addr))
+		      continue; /* skip this lease because another static address is configured */
+
 		    add_address(state, c, c->lease_time, NULL, &min_time, &req_addr, now);
 		    mark_context_used(state, &req_addr);
 		    get_context_tag(state, c);
@@ -860,7 +874,7 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 			put_opt6_string(_("address unavailable"));
 			end_opt6(o1);
 		      }
-		    else if (!check_address(state, &req_addr))
+		    else if (!check_address(state, config, &req_addr))
 		      {
 			/* Address leased to another DUID/IAID */
 			o1 = new_opt6(OPTION6_STATUS_CODE);
@@ -989,7 +1003,16 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 		    (this_context = address6_valid(state->context, &req_addr, tagif, 1)))
 		  {
 		    unsigned int lease_time;
-
+		    /* check if another static address is preferred */
+		    if (state->mac_len && config &&
+		        config_has_mac(config, state->mac, state->mac_len, state->mac_type) &&
+		        !config_implies(config, this_context, &req_addr))
+		      {
+			preferred_time = valid_time = 0;
+			message = _("deprecated");
+		      }
+		    else
+		      {
 		    get_context_tag(state, this_context);
 		    
 		    if (config_implies(config, this_context, &req_addr) && have_config(config, CONFIG_TIME))
@@ -1015,7 +1038,7 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 		    
 		    if (preferred_time == 0)
 		      message = _("deprecated");
-
+		        }
 		    address_assigned = 1;
 		  }
 		else
@@ -1071,11 +1094,22 @@ static int dhcp6_no_relay(struct state *state, int msg_type, unsigned char *inbu
 		 ia_option = opt6_find(opt6_next(ia_option, ia_end), ia_end, OPTION6_IAADDR, 24))
 	      {
 		struct in6_addr req_addr;
+		struct dhcp_context *c;
+		int config_addr_ok = 1;
 
 		/* alignment */
 		memcpy(&req_addr, opt6_ptr(ia_option, 0), IN6ADDRSZ);
-		
-		if (!address6_valid(state->context, &req_addr, tagif, 1))
+
+		c = address6_valid(state->context, &req_addr, tagif, 1);
+
+		if (c && state->mac_len && config &&
+		    config_has_mac(config, state->mac, state->mac_len, state->mac_type) &&
+		    !config_implies(config, c, &req_addr))
+		  {
+		    config_addr_ok = 0;
+		  }		
+
+		if (!c || !config_addr_ok)
 		  {
 		    o1 = new_opt6(OPTION6_STATUS_CODE);
 		    put_opt6_short(DHCP6NOTONLINK);
@@ -1703,6 +1737,10 @@ static int check_address(struct state *state, struct in6_addr *addr)
 { 
   struct dhcp_lease *lease;
 
+  if (state->mac_len && config &&
+      config_has_mac(config, state->mac, state->mac_len, state->mac_type))
+    return 1;
+
   if (!(lease = lease6_find_by_addr(addr, 128, 0)))
     return 1;
 
@@ -1779,7 +1817,7 @@ static int config_valid(struct dhcp_config *config, struct dhcp_context *context
 	  {
 	    setaddr6part(addr, addrpart+i);
 	    
-	    if (check_address(state, addr))
+	    if (check_address(state, config, addr))
 	      return 1;
 	  }
       }
-- 
2.17.1

