From 24574fd7f67c50d22428808c39666d504818fd2c Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Mon, 25 Nov 2024 02:11:41 -0500
Subject: [PATCH] backtrace module info

---
 lib/vsprintf.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/lib/vsprintf.c b/lib/vsprintf.c
index c5e2ec9..8f3e327 100644
--- a/lib/vsprintf.c
+++ b/lib/vsprintf.c
@@ -983,8 +983,11 @@ char *symbol_string(char *buf, char *end, void *ptr,
 		    struct printf_spec spec, const char *fmt)
 {
 	unsigned long value;
-#ifdef CONFIG_KALLSYMS
 	char sym[KSYM_SYMBOL_LEN];
+#ifndef CONFIG_KALLSYMS
+	struct module *mod;
+	int len;
+	struct module_memory *core_text;
 #endif
 
 	if (fmt[1] == 'R')
@@ -1005,8 +1008,15 @@ char *symbol_string(char *buf, char *end, void *ptr,
 
 	return string_nocheck(buf, end, sym, spec);
 #else
-	return special_hex_number(buf, end, value, sizeof(void *));
+	len = snprintf(sym, sizeof(sym), "0x%lx", value);
+	mod = __module_address(value);
+	core_text = &mod->mem[MOD_TEXT];
+	if (mod)
+		snprintf(sym + len, sizeof(sym) - len, " [%s@%p+0x%x]",
+			 mod->name, core_text->base,
+			 core_text->size);
 #endif
+	return string(buf, end, sym, spec);
 }
 
 static const struct printf_spec default_str_spec = {
-- 
2.17.1

