From 17e1a8472101baaa671dff672a5d94318965454d Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Sun, 28 May 2023 15:21:58 +0900
Subject: [PATCH] fix compilation problems for kernels other than 5.4

---
 nss_meminfo.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/nss_meminfo.c b/nss_meminfo.c
index 29bc594..d78470f 100644
--- a/nss_meminfo.c
+++ b/nss_meminfo.c
@@ -687,8 +687,11 @@ bool nss_meminfo_init(struct nss_ctx_instance *nss_ctx)
 	/*
 	 * meminfo_start is the label where the start address of meminfo map is stored.
 	 */
-	meminfo_start = (uint32_t *)ioremap_nocache(nss_ctx->load + NSS_MEMINFO_MAP_START_OFFSET,
-							NSS_MEMINFO_RESERVE_AREA_SIZE);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	meminfo_start =ioremap(nss_ctx->load + NSS_MEMINFO_MAP_START_OFFSET,NSS_MEMINFO_RESERVE_AREA_SIZE);
+#else
+	meminfo_start = (uint32_t *)ioremap_nocache(nss_ctx->load + NSS_MEMINFO_MAP_START_OFFSET,NSS_MEMINFO_RESERVE_AREA_SIZE);
+#endif
 	if (!meminfo_start) {
 		nss_info_always("%p: cannot remap meminfo start\n", nss_ctx);
 		return false;
--- a/nss_ppe.c
+++ b/nss_ppe.c
@@ -354,7 +354,11 @@ void nss_ppe_init(void)
 	/*
 	 * Get the PPE base address
 	 */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	ppe_pvt.ppe_base = ioremap(PPE_BASE_ADDR, PPE_REG_SIZE);
+#else
 	ppe_pvt.ppe_base = ioremap_nocache(PPE_BASE_ADDR, PPE_REG_SIZE);
+#endif
 	if (!ppe_pvt.ppe_base) {
 		nss_warning("DRV can't get PPE base address\n");
 		return;
--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -78,7 +78,11 @@ int nss_hal_firmware_load(struct nss_ctx_instance *nss_ctx, struct platform_devi
 		return rc;
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	load_mem = ioremap(npd->load_addr, nss_fw->size);
+#else
 	load_mem = ioremap_nocache(npd->load_addr, nss_fw->size);
+#endif
 	if (!load_mem) {
 		nss_info_always("%p: ioremap_nocache failed: %x", nss_ctx, npd->load_addr);
 		release_firmware(nss_fw);
--- a/nss_data_plane/nss_data_plane_edma.c
+++ b/nss_data_plane/nss_data_plane_edma.c
@@ -19,7 +19,7 @@
 #include "nss_tx_rx_common.h"
 #include <nss_dp_api_if.h>
 
-#define NSS_DP_EDMA_SUPPORTED_FEATURES (NETIF_F_HIGHDMA | NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_SG | NETIF_F_FRAGLIST | (NETIF_F_TSO | NETIF_F_TSO6 | NETIF_F_UFO))
+#define NSS_DP_EDMA_SUPPORTED_FEATURES (NETIF_F_HIGHDMA | NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_SG | NETIF_F_FRAGLIST | (NETIF_F_TSO | NETIF_F_TSO6))
 #define NSS_DATA_PLANE_EDMA_MAX_INTERFACES 6
 #define NSS_DATA_PLANE_EDMA_MAX_MTU_SIZE 9216
 #define NSS_DATA_PLANE_EDMA_PREHEADER_SIZE 32
--- a/nss_data_plane/nss_data_plane_edma.c
+++ b/nss_data_plane/nss_data_plane_edma.c
@@ -266,7 +266,7 @@ static bool nss_data_plane_register_to_nss_dp(struct nss_ctx_instance *nss_ctx,
 	bool is_open;
 	int core;
 
-	netdev = nss_dp_get_netdev_by_macid(if_num);
+	netdev = nss_dp_get_netdev_by_nss_if_num(if_num);
 	if (!netdev) {
 		nss_info("%p: Platform don't have data plane%d enabled, don't bring up nss_phys_if and don't register to nss-dp\n", nss_ctx, if_num);
 		return false;
--- a/nss_hal/ipq60xx/nss_hal_pvt.c
+++ b/nss_hal/ipq60xx/nss_hal_pvt.c
@@ -192,14 +192,21 @@ static struct nss_platform_data *__nss_hal_of_get_pdata(struct platform_device *
 	 */
 	npd->nphys = res_nphys.start;
 	npd->qgic_phys = res_qgic_phys.start;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	npd->nmap = ioremap(npd->nphys, resource_size(&res_nphys));
+#else
 	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+#endif
 	if (!npd->nmap) {
 		nss_info_always("%p: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
 	}
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	npd->qgic_map = ioremap(npd->qgic_phys, resource_size(&res_qgic_phys));
+#else
 	npd->qgic_map = ioremap_nocache(npd->qgic_phys, resource_size(&res_qgic_phys));
+#endif
+
 	if (!npd->qgic_map) {
 		nss_info_always("%p: nss%d: ioremap() fail for qgic map\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -400,14 +407,22 @@ static int __nss_hal_common_reset(struct platform_device *nss_dev)
 	}
 
 	of_node_put(cmn);
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	nss_misc_reset = ioremap(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+#else
 	nss_misc_reset = ioremap_nocache(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+#endif
+
 	if (!nss_misc_reset) {
 		pr_err("%p: ioremap fail for nss_misc_reset\n", nss_dev);
 		return -EFAULT;
 	}
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	nss_misc_reset_flag = ioremap(res_nss_misc_reset_flag.start, resource_size(&res_nss_misc_reset_flag));
+#else
 	nss_misc_reset_flag = ioremap_nocache(res_nss_misc_reset_flag.start, resource_size(&res_nss_misc_reset_flag));
+#endif
+
 	if (!nss_misc_reset_flag) {
 		pr_err("%p: ioremap fail for nss_misc_reset_flag\n", nss_dev);
 		return -EFAULT;
diff --git a/nss_hal/ipq806x/nss_hal_pvt.c b/nss_hal/ipq806x/nss_hal_pvt.c
index 875fc47..bbd31d1 100644
--- a/nss_hal/ipq806x/nss_hal_pvt.c
+++ b/nss_hal/ipq806x/nss_hal_pvt.c
@@ -457,8 +457,11 @@ static struct nss_platform_data *__nss_hal_of_get_pdata(struct platform_device *
 	 */
 	npd->nphys = res_nphys.start;
 	npd->vphys = res_vphys.start;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	npd->nmap = ioremap(npd->nphys, resource_size(&res_nphys));
+#else
 	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+#endif
 	if (!npd->nmap) {
 		nss_info_always("%p: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -709,8 +712,12 @@ static int __nss_hal_common_reset(struct platform_device *nss_dev)
 		return -EFAULT;
 	}
 	of_node_put(cmn);
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	fpb_base = ioremap(res_nss_fpb_base.start, resource_size(&res_nss_fpb_base));
+#else
 	fpb_base = ioremap_nocache(res_nss_fpb_base.start, resource_size(&res_nss_fpb_base));
+#endif
+
 	if (!fpb_base) {
 		pr_err("%p: ioremap fail for nss_fpb_base\n", nss_dev);
 		return -EFAULT;
diff --git a/nss_hal/ipq807x/nss_hal_pvt.c b/nss_hal/ipq807x/nss_hal_pvt.c
index 3b09e07..8ca152c 100644
--- a/nss_hal/ipq807x/nss_hal_pvt.c
+++ b/nss_hal/ipq807x/nss_hal_pvt.c
@@ -232,8 +232,12 @@ static struct nss_platform_data *__nss_hal_of_get_pdata(struct platform_device *
 	npd->nphys = res_nphys.start;
 	npd->vphys = res_vphys.start;
 	npd->qgic_phys = res_qgic_phys.start;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	npd->nmap = ioremap(npd->nphys, resource_size(&res_nphys));
+#else
 	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+#endif
+
 	if (!npd->nmap) {
 		nss_info_always("%p: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -244,8 +248,11 @@ static struct nss_platform_data *__nss_hal_of_get_pdata(struct platform_device *
 		nss_info_always("%p: nss%d: ioremap() fail for vphys\n", nss_ctx, nss_ctx->id);
 		goto out;
 	}
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	npd->qgic_map = ioremap(npd->qgic_phys, resource_size(&res_qgic_phys));
+#else
 	npd->qgic_map = ioremap_nocache(npd->qgic_phys, resource_size(&res_qgic_phys));
+#endif
 	if (!npd->qgic_map) {
 		nss_info_always("%p: nss%d: ioremap() fail for qgic map\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -462,8 +469,11 @@ static int __nss_hal_common_reset(struct platform_device *nss_dev)
 		return -EFAULT;
 	}
 	of_node_put(cmn);
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
+	nss_misc_reset = ioremap(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+#else
 	nss_misc_reset = ioremap_nocache(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+#endif
 	if (!nss_misc_reset) {
 		pr_err("%p: ioremap fail for nss_misc_reset\n", nss_dev);
 		return -EFAULT;
-- 
2.17.1

