From b2c0886fdc73876a0fdf14bc8140edf4175ed32f Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Wed, 17 May 2023 23:24:20 -0400
Subject: [PATCH] change tmpfs default size to two-thirds

---
 mm/shmem.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mm/shmem.c b/mm/shmem.c
index aae2f40..860679d 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -125,14 +125,14 @@ struct shmem_options {
 #ifdef CONFIG_TMPFS
 static unsigned long shmem_default_max_blocks(void)
 {
-	return totalram_pages() / 2;
+	return div_u64(totalram_pages() * 2 , 3);
 }
 
 static unsigned long shmem_default_max_inodes(void)
 {
 	unsigned long nr_pages = totalram_pages();
 
-	return min(nr_pages - totalhigh_pages(), nr_pages / 2);
+	return min(nr_pages - totalhigh_pages(), div_u64(nr_pages * 2, 3));
 }
 #endif
 
-- 
2.17.1

