From bc04ef4827b4a797e5e4fb331b44e9f4c59febff Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Thu, 11 May 2023 05:11:24 -0400
Subject: [PATCH] 1

---
 include/poll.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/poll.h b/include/poll.h
index 24691f1..33d25e9 100644
--- a/include/poll.h
+++ b/include/poll.h
@@ -38,7 +38,7 @@ _FORTIFY_FN(poll) int poll(struct pollfd *__f, nfds_t __n, int __s)
 		__builtin_trap();
 	return __orig_poll(__f, __n, __s);
 }
-
+#if 0
 #ifdef _GNU_SOURCE
 #undef ppoll
 _FORTIFY_FN(ppoll) int ppoll(struct pollfd *__f, nfds_t __n, const struct timespec *__s,
@@ -51,7 +51,7 @@ _FORTIFY_FN(ppoll) int ppoll(struct pollfd *__f, nfds_t __n, const struct timesp
 	return __orig_ppoll(__f, __n, __s, __m);
 }
 #endif
-
+#endif
 #ifdef __cplusplus
 }
 #endif
-- 
2.17.1

