From 485def7c82bba0313224e268d472ded94adceb61 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Fri, 6 May 2022 05:38:23 -0400
Subject: [PATCH] add overlay option

---
 block.c | 228 +++++++++++++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 199 insertions(+), 29 deletions(-)

diff --git a/block.c b/block.c
index f5ec5cf..b36edce 100644
--- a/block.c
+++ b/block.c
@@ -671,27 +671,60 @@ static char* find_mount_point(char *block)
 	return point;
 }
 
-static int print_block_uci(struct probe_info *pr)
+static int print_block_uci(struct probe_info *pr,int fd)
 {
-	if (!strcmp(pr->type, "swap")) {
-		printf("config 'swap'\n");
-	} else {
-		char *mp = find_mount_point(pr->dev);
+	if(fd<0)
+	{
+		if (!strcmp(pr->type, "swap")) {
+			printf("config 'swap'\n");
+		} else {
+			char *mp = find_mount_point(pr->dev);
 
-		printf("config 'mount'\n");
-		if (mp) {
-			printf("\toption\ttarget\t'%s'\n", mp);
-			free(mp);
+			printf("config 'mount'\n");
+			if (mp) {
+				printf("\toption\ttarget\t'%s'\n", mp);
+				free(mp);
+			} else {
+				printf("\toption\ttarget\t'/mnt/%s'\n", basename(pr->dev));
+			}
+		}
+		if (pr->uuid)
+			printf("\toption\tuuid\t'%s'\n", pr->uuid);
+		else
+			printf("\toption\tdevice\t'%s'\n", pr->dev);
+		printf("\toption\tenabled\t'0'\n\n");
+	}else
+	{
+		char path[1024];
+		if (!strcmp(pr->type, "swap")) {
+			write(fd,"config 'swap'\n",strlen("config 'swap'\n"));
 		} else {
-			printf("\toption\ttarget\t'/mnt/%s'\n", basename(pr->dev));
+			char *mp = find_mount_point(pr->dev);
+
+			write(fd,"config 'mount'\n",strlen("config 'mount'\n"));
+			if (mp) {
+				snprintf(path, sizeof(path), "\toption\ttarget\t'%s'\n", mp);
+				write(fd,path,strlen(path));
+				free(mp);
+			} else {
+				snprintf(path, sizeof(path), "\toption\ttarget\t'/mnt/%s'\n", basename(pr->dev));
+				write(fd,path,strlen(path));
+			}
 		}
-	}
-	if (pr->uuid)
-		printf("\toption\tuuid\t'%s'\n", pr->uuid);
-	else
-		printf("\toption\tdevice\t'%s'\n", pr->dev);
-	printf("\toption\tenabled\t'0'\n\n");
+		if (pr->uuid)
+		{
+			snprintf(path, sizeof(path), "\toption\tuuid\t'%s'\n", pr->uuid);
+			write(fd,path,strlen(path));
 
+		}
+		else
+		{
+			snprintf(path, sizeof(path), "\toption\tdevice\t'%s'\n", pr->dev);
+			write(fd,path,strlen(path));
+		}
+		write(fd,"\toption\tenabled\t'0'\n\n",strlen("\toption\tenabled\t'0'\n\n"));
+	}
+	
 	return 0;
 }
 
@@ -1723,24 +1756,158 @@ static int main_umount(int argc, char **argv)
 	return 0;
 }
 
-static int main_detect(int argc, char **argv)
+static int main_detect(int argc, char **argv,int file)
 {
 	struct probe_info *pr;
 
 	cache_load(0);
-	printf("config 'global'\n");
-	printf("\toption\tanon_swap\t'0'\n");
-	printf("\toption\tanon_mount\t'1'\n");
-	printf("\toption\tauto_swap\t'1'\n");
-	printf("\toption\tauto_mount\t'1'\n");
-	printf("\toption\tdelay_root\t'5'\n");
-	printf("\toption\tcheck_fs\t'0'\n\n");
-	list_for_each_entry(pr, &devices, list)
-		print_block_uci(pr);
-
+	if(!file)
+	{
+		printf("config 'global'\n");
+		printf("\toption\tanon_swap\t'0'\n");
+		printf("\toption\tanon_mount\t'1'\n");
+		printf("\toption\tauto_swap\t'1'\n");
+		printf("\toption\tauto_mount\t'1'\n");
+		printf("\toption\tdelay_root\t'5'\n");
+		printf("\toption\tcheck_fs\t'0'\n\n");
+		list_for_each_entry(pr, &devices, list)
+			print_block_uci(pr,-1);
+	}else
+	{
+		int fd=-1; 
+		if ((fd= open((char*)argv, O_RDWR|O_CREAT|O_TRUNC, 0666)) >= 0)
+		{
+			write(fd,"config 'global'\n",strlen("config 'global'\n"));
+			write(fd,"\toption\tanon_swap\t'0'\n",strlen("\toption\tanon_swap\t'0'\n"));
+			write(fd,"\toption\tanon_mount\t'1'\n",strlen("\toption\tanon_mount\t'1'\n"));
+			write(fd,"\toption\tauto_swap\t'1'\n",strlen("\toption\tauto_swap\t'1'\n"));
+			write(fd,"\toption\tauto_mount\t'1'\n",strlen("\toption\tauto_mount\t'1'\n"));
+			write(fd,"\toption\tdelay_root\t'5'\n",strlen("\toption\tdelay_root\t'5'\n"));
+			write(fd,"\toption\tcheck_fs\t'0'\n\n",strlen("\toption\tcheck_fs\t'0'\n\n"));
+			list_for_each_entry(pr, &devices, list)
+				print_block_uci(pr,fd);
+			close(fd);
+		}
+	}
 	return 0;
 }
 
+static int main_overlay(int argc, char **argv)
+{
+	char path[64];
+	struct probe_info *pr;
+	char blkdev_path[32] = { 0 };
+#ifdef UBIFS_EXTROOT
+	libubi_t libubi;
+#endif
+	find_block_mtd("\"rootfs_data\"", blkdev_path, sizeof(blkdev_path));
+	if (blkdev_path[0]) {
+		pr = find_block_info(NULL, NULL, blkdev_path);
+		if (pr && !strcmp(pr->type, "jffs2")) {
+			char cfg[] = "/tmp/jffs_cfg";
+
+			/*
+			 * Mount MTD part and try extroot (using
+			 * /etc/config/fstab from that partition)
+			 */
+			mkdir_p(cfg, 0755);
+			if (!mount(blkdev_path, cfg, "jffs2", MS_NOATIME, NULL)) {
+				snprintf(path, sizeof(path), "%s/upper/etc/config/fstab", cfg);
+				if (access(path, R_OK) < 0)
+				{
+					snprintf(path, sizeof(path), "%s/etc/config/fstab", cfg);
+					if (access(path, R_OK) < 0)
+						memset(path,0,sizeof(path));
+				}
+				if(strlen(path))
+					main_detect(1,(char**)path,1);
+				umount2(cfg, MNT_DETACH);
+			}
+			rmdir(cfg);
+			return 0;
+		}
+	}
+#ifdef UBIFS_EXTROOT
+	/* ... but it also could be an UBI volume */
+	memset(blkdev_path, 0, sizeof(blkdev_path));
+	libubi = libubi_open();
+	find_block_ubi(libubi, "rootfs_data", blkdev_path, sizeof(blkdev_path));
+	libubi_close(libubi);
+	if (blkdev_path[0]) {
+		char cfg[] = "/tmp/ubifs_cfg";
+
+		/* Mount volume and try extroot (using fstab from that vol) */
+		mkdir_p(cfg, 0755);
+		if (!mount(blkdev_path, cfg, "ubifs", MS_NOATIME, NULL)) {
+			snprintf(path, sizeof(path), "%s/upper/etc/config/fstab", cfg);
+			if (access(path, R_OK) < 0)
+			{
+				snprintf(path, sizeof(path), "%s/etc/config/fstab", cfg);
+				if (access(path, R_OK) < 0)
+					memset(path,0,sizeof(path));
+			}
+			if(strlen(path))
+				main_detect(1,(char**)path,1);
+			umount2(cfg, MNT_DETACH);
+		}
+		rmdir(cfg);
+		return 0;
+       }
+#endif
+
+  /* Find volume using libfstools */
+	struct volume *data = volume_find("rootfs_data");
+	if (data) {
+		volume_init(data);
+
+	  switch (volume_identify(data)) {
+		case FS_EXT4: {
+			char cfg[] = "/tmp/ext4_cfg";
+
+			/* Mount volume and try extroot (using fstab from that vol) */
+			mkdir_p(cfg, 0755);
+			if (!mount(data->blk, cfg, "ext4", MS_NOATIME, NULL)) {
+				snprintf(path, sizeof(path), "%s/upper/etc/config/fstab", cfg);
+				if (access(path, R_OK) < 0)
+				{
+					snprintf(path, sizeof(path), "%s/etc/config/fstab", cfg);
+					if (access(path, R_OK) < 0)
+						memset(path,0,sizeof(path));
+				}
+				if(strlen(path))
+					main_detect(1,(char**)path,1);
+
+				umount2(cfg, MNT_DETACH);
+			}
+			rmdir(cfg);
+			return 0;
+		}
+
+		case FS_F2FS: {
+			char cfg[] = "/tmp/f2fs_cfg";
+
+			/* Mount volume and try extroot (using fstab from that vol) */
+			mkdir_p(cfg, 0755);
+			if (!mount(data->blk, cfg, "f2fs", MS_NOATIME, NULL)) {
+				snprintf(path, sizeof(path), "%s/upper/etc/config/fstab", cfg);
+				if (access(path, R_OK) < 0)
+				{
+					snprintf(path, sizeof(path), "%s/etc/config/fstab", cfg);
+					if (access(path, R_OK) < 0)
+						memset(path,0,sizeof(path));
+				}
+				if(strlen(path))
+					main_detect(1,(char**)path,1);
+
+				umount2(cfg, MNT_DETACH);
+			}
+			rmdir(cfg);
+			return 0;
+		}
+	}
+	}
+return 0;
+}
 static int main_info(int argc, char **argv)
 {
 	int i;
@@ -1914,7 +2081,10 @@ int main(int argc, char **argv)
 			return main_info(argc, argv);
 
 		if (!strcmp(argv[1], "detect"))
-			return main_detect(argc, argv);
+			return main_detect(argc, argv,0);
+
+		if (!strcmp(argv[1], "overlay"))
+			return main_overlay(argc, argv);
 
 		if (!strcmp(argv[1], "hotplug"))
 			return main_hotplug(argc, argv);
@@ -1940,7 +2110,7 @@ int main(int argc, char **argv)
 		}
 	}
 
-	ULOG_ERR("Usage: block <info|mount|umount|detect>\n");
+	ULOG_ERR("Usage: block <info|mount|umount|detect|overlay>\n");
 
 	return -1;
 }
-- 
2.17.1

