From d987fb6647ef6c12e6a37af9e6d476be78d30168 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Fri, 1 Apr 2022 22:40:28 +0900
Subject: [PATCH] add link pthread

---
 scripts/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/scripts/Makefile
+++ b/scripts/Makefile
@@ -3,7 +3,7 @@
 # scripts contains sources for various helper programs used throughout
 # the kernel for the build process.
 
-CRYPTO_LIBS = $(shell $(HOSTPKG_CONFIG) --libs libcrypto 2> /dev/null || echo -lcrypto)
+CRYPTO_LIBS = $(shell $(HOSTPKG_CONFIG) --libs libcrypto 2> /dev/null || echo -lcrypto) -lpthread
 CRYPTO_CFLAGS = $(shell $(HOSTPKG_CONFIG) --cflags libcrypto 2> /dev/null)
 
 hostprogs-always-$(CONFIG_BUILD_BIN2C)			+= bin2c
