From 64e213b2f6f4e26b6ffb9bc128e408ae116dd043 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Wed, 20 Mar 2024 03:30:30 -0400
Subject: [PATCH] fix type error

---
 core/rtw_bt_mp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/core/rtw_bt_mp.c b/core/rtw_bt_mp.c
index 9f4f563..c0843bb 100644
--- a/core/rtw_bt_mp.c
+++ b/core/rtw_bt_mp.c
@@ -169,7 +169,7 @@ mptbt_CheckBtRspStatus(
     PBT_EXT_C2H			pExtC2h
 )
 {
-	BT_CTRL_STATUS	retStatus=BT_OP_STATUS_SUCCESS;
+	BT_CTRL_STATUS	retStatus=BT_STATUS_SUCCESS;
 
 	switch(pExtC2h->statusCode) {
 	case BT_OP_STATUS_SUCCESS:
-- 
2.17.1

