From 20e743bc3749d9f8cf29a1aca6df2951b0fc779e Mon Sep 17 00:00:00 2001
From: ame <383152993@qq.com>
Date: Thu, 27 Apr 2023 03:14:26 -0700
Subject: [PATCH] 2

---
 src/util/snprintf.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/util/snprintf.h b/src/util/snprintf.h
index 5a61554..2cb7280 100644
--- a/src/util/snprintf.h
+++ b/src/util/snprintf.h
@@ -53,7 +53,6 @@ upx_rsize_t upx_safe_strlen(const char *);
 #undef sprintf
 #undef vsnprintf
 #define snprintf upx_safe_snprintf
-#define sprintf ERROR_sprintf_IS_DANGEROUS_USE_snprintf
 #define vsnprintf upx_safe_vsnprintf
 
 #undef strlen
-- 
2.30.1 (Apple Git-130)

